### ISSUE TITLE
"Permission denied" when updating pods

### ISSUE THREAD

**Issue by dev-alltrails** [Appboy/appboy-ios-sdk#186]
This morning I attempted to update pods and received several "Permission denied" errors in the `FacebookSDKApplicationTests` library.  I tracked this down to the Braze SDK.

If I include this in the Podfile...

`	pod 'Appboy-iOS-SDK'`

...then I see many errors like this:

```
sh: line 66: src/FacebookSDKApplicationTests/ReferenceImages/FBLikeControlTests/testStyleStandard_0_0_0_2.png: Permission denied
sh: line 67: src/FacebookSDKApplicationTests/ReferenceImages/FBLikeControlTests/testStyleStandard_0_123_1_2.png: Permission denied
sh: line 68: src/FacebookSDKApplicationTests/ReferenceImages/FBLikeControlTests/testStyleStandard_0_0_2_1.png: Permission denied
sh: line 69: src/FacebookSDKApplicationTests/ReferenceImages/FBLikeControlTests/testStyleStandard_1_0_1_0.png: Permission denied
```

If I remove the Braze pod then everything works fine.

I have reproduced this issue on Cocoapods 1.6 and 1.7 prerelease.  We tried this on a second machine and saw the same issue.

**Comment by Bucimis** [comment]
@dev-alltrails thanks for filing - this is unexpected on our end as our SDK doesn't reference the FB SDK and has no similarly named files or anything else that would seem to cause something like this. Could you try clearing your cocoapods caches using `sudo rm -R  ~/Library/Caches/CocoaPods; sudo rm -R  ~/.cocoapods/repos` and trying again?

Could you also provide full steps to repro - especially the command you used to generate the errors as well as the Pods that are conflicting with ours?

**Comment by jasorod** [comment]
Hi, I'm seeing the same problem.  For some reason an ancient version of the Facebook-iOS-SDK (3.16.2) is getting downloaded whenever I include "Appboy-iOS-SDK" in my podfile.  The workaround I found was to start from scratch with a podfile that just contained the "Appboy-iOS-SDK" pod, install that, and then add back in the rest of my pods and install them.

**Comment by Bucimis** [comment]
Hi @jasorod, is there a minimal set of dependencies that causes the error to repro? we'd be interested in trying to repro on our end as well

**Comment by matheda** [comment]
@Bucimis attached is a zip file that contains the Podfile, Podfile.lock, and a verbose log of running "pod update".  

This afternoon I tried completely uninstalling Cocoapods and its various folders and reinstalling from scratch.  I tried with both the current release and the pre-release of Cocoapods.

I reduced the Podfile to just `pod 'Appboy-iOS-SDK'`.  I then added all of the other pods in one by one, and when I got to the end I received no error... very odd, but I thought I was in the clear.  Unfortunately there were many compiler warnings from several unrelated pods.  I removed `pod 'Appboy-iOS-SDK'` from the Podfile and ran pod update again; no errors and everything compiled.

[Permission Denied.zip](https://github.com/Appboy/appboy-ios-sdk/files/3113961/Permission.Denied.zip)

**Comment by matheda** [comment]
@Bucimis @jasorod I think I have found the cause of this issue.  There is some kind of dependency resolution issue introduced by Cocoapods when the SDWebImage pod is installed along with the Braze pod, but only when Podfile.lock is missing.  Skip to the end if you just want to see the workaround.

**Test 1**:  Podfile with `pod 'Appboy-iOS-SDK'` only: works fine

**Test 2**: Podfile with `pod 'SDWebImage'` only: works fine

**Test 3**: Podfile with both `pod 'Appboy-iOS-SDK'` and `pod 'SDWebImage'`: works fine

**Test 4**: Repeat Test 3, but remove Podfile.lock first.  You will find that it sits on "Analyzing dependencies" for several minutes (odd for just 2 pods) and then produces the permission denied error when attempting to install the Facebook SDK:

```
Installing Appboy-iOS-SDK (2.9.6)
Installing Bolts (1.9.0)
Installing Facebook-iOS-SDK (3.16.2)
[!] /bin/bash -c 
set -e
find src -name \*.png | grep -v @ | grep -v -- - | sed -e 's|\(.*\)/\([a-zA-Z0-9]*\).png|scripts/image_to_code.py -i \1/\2.png -c \2 -o src|' | sh && find src -name \*.wav | grep -v @ | grep -v -- - | sed -e 's|\(.*\)/\([a-zA-Z0-9]*\).wav|scripts/audio_to_code.py -i \1/\2.wav -c \2 -o src|'  | sh

sh: line 12: src/FacebookSDKApplicationTests/ReferenceImages/FBLikeControlTests/testStyleBoxCount_1_0_2.png: Permission denied

```

So here is the odd part:  Neither Braze nor SDWebImage seem to link to the Facebook SDK.  If I delete Podfile.lock again and install them individually I see the following dependencies for Braze...

```
Using Appboy-iOS-SDK (3.14.0)
Using FLAnimatedImage (1.0.12)
Using SDWebImage (4.4.6)

```

... and no dependencies for SDWebImage:

```
Installing SDWebImage (5.0.1)
```

However, if they are both present, then the Facebook pod mysteriously shows up.  Also note that for some reason it causes the Braze pod to downgrade from 3.14.0 to 2.9.6.  Also note that when Braze installs by itself, it has a dependency to SDWebImage but installs 4.4.6 rather than 5.0.1.

It looks like Braze is linking to an older version of SDWebImage, so I deleted Podfile.lock and changed my Podfile to...

```
pod 'Appboy-iOS-SDK'
pod 'SDWebImage', '4.4.6'
```

... and everything worked fine:

```
Using Appboy-iOS-SDK (3.14.0)
Using FLAnimatedImage (1.0.12)
Using SDWebImage (4.4.6)
Generating Pods project
```

I suspect that this is an issue with the dependency resolver in Cocoapods rather than an issue with the Braze SDK but would like to get confirmation before opening a ticket with them.

**Comment by Bucimis** [comment]
@matheda Thanks for taking the time to do this extensive investigation, its very much appreciated. I have a few thoughts and did take one action:

1) We currently link against SDWebImage 4.4.6 and your final set of deps looks correct. We're planning to add 5.x support in the near future but don't have an eta yet.
2) Braze 2.9.6 did explicitly depend on the Facebook SDK - see https://github.com/Appboy/appboy-ios-sdk/blob/b3d7182bd632d89773a71aa745f35b8e9e36e9d5/Appboy-iOS-SDK.podspec#L34. When Cocoapods tried to downgrade the Braze SDK to 2.9.6, it's not surprising that it also tried to pull in the FB dependency.
3) Braze 2.9.6 was the only version of the SDK to depend on the Facebook SDK explicitly in its podspec and was also the only version where the SDWebImage dependency was declared in the AppboyKit subspec instead of the top level spec (at least among nearby releases, I didn't check all). I suspect there was a perverse dependency resolution situation where Cocoapods was trying to find the most recent version of Appboy-iOS-SDK that didn't depend on SDWebImage to live alongside the newer SDWebImage (5.x) version, and ended up choosing the super old/obsolete 2.9.6 SDK that didn't declare SDWebImage in the top level spec.
4) I've gone ahead and deleted the problematic 2.9.6 spec from Cocoapods. I'm hoping this resolves things for most folks - if it ends up causing Cocoapods to problematically add even older Braze dependencies, we'll probably just have to keep this in mind as a known issue until we're able to upgrade SDWebImage to 5.x.

Let me know your thoughts on ^ and thanks again.

**Comment by matheda** [comment]
Glad to help, and I think you have filled in the part that I was scratching my head over.  I think you are exactly right on what Cocoapods is doing, and it's probably not wrong, it just worked out odd in this case.  Thanks for the additional information.  I was the original submitter of this ticket (posting as dev-alltrails), so feel free to close this ticket.

**Comment by Bucimis** [comment]
@matheda yeah definitely an unusual case. Thanks again & let us know if anything else comes up!

**Comment by Bucimis** [comment]
Closing the loop here - on Friday we released 3.15.0 which updates our SDWebImage major version dependency from 4 to 5. thanks all for your participation!
